Fix for `add_ancillary_variable` (#2820)

jlenh opened the pull request:
## Description

As described in #2820, in some cases where the supplementary variables and the dataset variables have slightly different values for the coordinates, the direct matching done via iris might fail in
https://github.com/ESMValGroup/ESMValCore/blob/05c77e9e54f658c447ebe979946c45266cb8174b/esmvalcore/preprocessor/_supplementary_vars.py#L148-L160

To avoid this, this PR includes additional checks to match the correct coordinate in both input cubes:
- Direct check via iris as described previously
- Try casting the coordinates back and forth between np.float32 and np.float64 in case it was just a casting issue when writing the data
- Try finding if the coordinate points are "close" using the an absolute tolerance threshold close to the difference in the actual data.
If nothing matches then an iris.exceptions.CoordinateNotFoundError is returned.

Note: I had to bypass the erros flagged by ruff for C901, PLR0912, PLR0915, eg. too complex, too many branches, and too many statements. Happy to get some feedback on how I could address these. This is a first try to fix this issue.

Closes #2820 

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [ ] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [ ] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [ ] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [ ] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [ ] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [ ] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [ ] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [ ] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [ ] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

bouweandela reviewed (changes requested):
>  had to bypass the erros flagged by ruff for C901, PLR0912, PLR0915, eg. too complex, too many branches, and too many statements. Happy to get some feedback on how I could address these.

You could move the code between line 141 and 247 into a separate function, called e.g. `get_data_dims`

bouweandela commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -146,28 +147,104 @@ def add_ancillary_variable(
+                                    ),
+                                ).flatten()
+                                diffs = np.concatenate((dy, dx))
+                                diffs = diffs[diffs > 0]
```
I believe that only works if the data is on a grid with rectangular cells. That may not always be the case, e.g. would this work on the [ICON grid](https://www.dwd.de/EN/research/weatherforecasting/num_modelling/01_num_weather_prediction_modells/icon_description.html)?

I also tried, just to convince myself, with a random ocean model (`CMIP6/ScenarioMIP/NCAR/CESM2/ssp585/r10i1p1f1/Omon/tos/gn/v20200528/tos_Omon_CESM2_ssp585_r10i1p1f1_gn_206501-210012.nc`), and got a value of 3e-8 for `min_diff`, which seems smaller than what we are looking for here.

jlenh replied:
Hmmm, I though that for the ICON grid, even though the structure is very different, the points are somewhat organised in "groups", and thus some (spatially) neighboring points are also neighbors in the unstructured coordinate.

As for the ocean example, it seems indeed not ideal to yield such a small difference..

Should we just skip higher-dimensional examples for now then? It seems overly complicated to get a general solution when this is only relevant for a handful of (older) model outputs so far (and hopefully it shouldn't happen for newer versions of models)

jlenh commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -146,28 +147,104 @@ def add_ancillary_variable(
+                    for cube_coord in cube.coords():
+                        if (
+                            cube_coord.dtype == np.float64
+                            and coord.standard_name == cube_coord.standard_name
```
Sounds good, I'll add that in

jlenh commented:
Following Bouwe's comments, I have updated a few things (still not at 100% for the test coverage though I think).

For example I kept the initial recommendation from https://github.com/ESMValGroup/ESMValCore/pull/2825#discussion_r2313417269 because the `metadata` check is exactly what happens under the hood of `cube.coord_dims(coord)` so if it failed beforehand then it would fail here again. I also had to adapt this part to make sure the preprocessor works for `recipe_russel18jgr.yml` which was failing because of that (now there is an NCL error but that is another topic).

bouweandela commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +104,123 @@ def add_cell_measure(
+                    if (
+                        np.issubdtype(cube_coord.dtype, np.number)
+                        and (
+                            cube_coord.standard_name == coord.standard_name
+                            or cube_coord.var_name == coord.var_name
```
```suggestion
                            cube_coord.var_name == coord.var_name
                            or (coord.standard_name is not None and cube_coord.standard_name == coord.standard_name)
```

bouweandela commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +104,123 @@ def add_cell_measure(
+                        and cube_coord.points.shape == coord.points.shape
+                    ):
+                        # Trying to cast back and forth the coordinate points
+                        logger.debug("Trying to cast coordinates...")
```
```suggestion
```

bouweandela commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +104,123 @@ def add_cell_measure(
+                            coord_.core_points()
+                            .astype(np.float32)
+                            .astype(np.float64)
+                        )
+                        if np.all(coord_.points == cube_coord.points):
```
```suggestion
                        )
                        cube_coord = cube_coord.copy()
                        if np.all(coord_.points == cube_coord.points):
```
To avoid modifying the input cube by realizing its coordinate data on the next line

bouweandela commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +104,123 @@ def add_cell_measure(
+                        )
+                        if np.all(coord_.points == cube_coord.points):
+                            cube_dims = cube_c.coord_dims(cube_coord)
+                            logger.debug("Found a matching casted coordinate.")
```
If you want to keep the log message, I would recommend making it more descriptive, e.g. mention exactly which coordinate matches on what cube. This will make the log message useful, even when running with `max_parallel_tasks > 1`, when log messages from multiple preprocessor functions running in parallel all end up in the same log file.

jlenh replied:
I see! I will update that.

bouweandela commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +104,123 @@ def add_cell_measure(
+                            break
+                        # Trying to find a "close" coordinate for 1D cases
+                        if cube_coord.points.ndim == 1:
+                            logger.debug(
+                                "Trying to find a close match in the cube coords...",
+                            )
```
```suggestion
```

bouweandela commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +104,123 @@ def add_cell_measure(
+                                atol=atol,
+                            ):
+                                cube_dims = cube_c.coord_dims(cube_coord)
+                                logger.debug("Found a close coordinate.")
```
Same as above, delete or make more descriptive

bouweandela commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +104,123 @@ def add_cell_measure(
+                for ancillary_dim, cube_dim in zip(
+                    cube_a.coord_dims(coord),
+                    cube_dims,
+                    strict=False,
```
```suggestion
                    strict=True,
```

jlenh replied:
Updated in https://github.com/ESMValGroup/ESMValCore/pull/2825/commits/6432ef3ac8dba509622e6ed32c0fabd928b93308

bouweandela commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +104,123 @@ def add_cell_measure(
+                logger.debug(
+                    "No metadata match from ancillary coord in cube coords.",
+                )
+                cube_dims = []
```
```suggestion
                cube_dims = None
```
Should the default value be `None` here, to signal that no coordinate was found on line 209?

bouweandela commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +104,123 @@ def add_cell_measure(
+                        and cube_coord.points.shape == coord.points.shape
+                    ):
+                        # Trying to cast back and forth the coordinate points
+                        logger.debug("Trying to cast coordinates...")
+                        coord_ = coord.copy()
+                        coord_.points = (
+                            coord_.core_points()
+                            .astype(np.float32)
+                            .astype(np.float64)
+                        )
+                        if np.all(coord_.points == cube_coord.points):
+                            cube_dims = cube_c.coord_dims(cube_coord)
+                            logger.debug("Found a matching casted coordinate.")
+                            break
+                        # Trying to find a "close" coordinate for 1D cases
+                        if cube_coord.points.ndim == 1:
+                            logger.debug(
+                                "Trying to find a close match in the cube coords...",
+                            )
+                            cube_coord.points = (
+                                cube_coord.core_points()
+                                .astype(np.float32)
+                                .astype(np.float64)
+                            )
+                            acceptable_relative_difference = 0.1
+                            min_diff = np.abs(
+                                np.diff(
+                                    cube_coord.core_points(),
+                                ),
+                            ).min()
+                            atol = acceptable_relative_difference * min_diff
+                            if np.allclose(
+                                coord.points,
+                                cube_coord.core_points(),
+                                rtol=0,
+                                atol=atol,
+                            ):
+                                cube_dims = cube_c.coord_dims(cube_coord)
+                                logger.debug("Found a close coordinate.")
+                                break
```
Move this to a function called e.g. `find_matching_coord` to reduce complexity

bouweandela commented on tests/integration/preprocessor/_supplementary_vars/test_add_supplementary_variables.py:
```diff
@@ -281,6 +282,77 @@ def test_add_ancillary_vars_errors(self):
+                ancillary_dummy,
+                ancillary_dummy_cube,
+            )
+        # Ancillary var has only one dimension match
+        ancillary_dummy = iris.cube.Cube(
+            np.ones((3, 3)),
+            dim_coords_and_dims=[(self.lats, 0), (plev_dim, 1)],
+        )
+        ancillary_dummy_cube = iris.coords.AncillaryVariable(
+            ancillary_dummy.core_data(),
+            standard_name=ancillary_dummy.standard_name,
+            units=ancillary_dummy.units,
+            var_name=ancillary_dummy.var_name,
+            attributes=ancillary_dummy.attributes,
+            long_name=ancillary_dummy.long_name,
+        )
+        with pytest.raises(iris.exceptions.CoordinateNotFoundError):
+            _ = get_data_dims(
+                cube,
+                ancillary_dummy,
+                ancillary_dummy_cube,
+            )
```
This looks like a new test, please move it to a new `test_...` function (you can put the shared code in a [fixture](https://docs.pytest.org/en/6.2.x/fixture.html#back-to-fixtures)) or use a [parametrized test](https://docs.pytest.org/en/stable/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions) to generate separate tests for each scenario you are testing

bouweandela reviewed (changes requested):
Thank you for the improvements! Could you also add test cases for matching the coordinates when the coordinate data is almost but not quite equal?

jlenh commented:
> Thank you for the improvements! Could you also add test cases for matching the coordinates when the coordinate data is almost but not quite equal?

I added these tests for both functions `get_data_dims` and `find_matching_coord` on top of addressing your comments!

Test coverage is at 95%, the bit of code that is flagged should be covered by the tests of https://github.com/ESMValGroup/ESMValCore/pull/2825/commits/aa6dd2a29149a8272d70b04de74de742ea956dac I think.

bouweandela commented:
Looking at why your test isn't running the code we expected it would run, I'm afraid we may have been on the wrong track with this. The [documentation for iris.cube.Cube.dim_coords`](https://scitools-iris.readthedocs.io/en/stable/generated/api/iris.cube.html#iris.cube.Cube.coord_dims) says:

> Return a tuple of the data dimensions relevant to the given coordinate.
>
> When searching for the given coordinate in the cube the comparison is made using coordinate metadata equality. Hence the given coordinate instance need not exist on the cube, and may contain different coordinate values.

I tried this out and indeed the coordinate values are completely ignored:

```python
In [1]: from iris.cube import Cube

In [2]: from iris.coords import DimCoord

In [3]: cube = Cube([0], dim_coords_and_dims=[(DimCoord([1], var_name='lat'), 0),])

In [4]: coord = DimCoord([90], var_name='lat')

In [5]: cube.coord_dims(coord)
Out[5]: (0,)
```

That makes it seem likely that the problem reported in https://github.com/ESMValGroup/ESMValCore/issues/2820 has been caused by differences in coordinate metadata all along. Perhaps a differing long name or differing attributes on the coordinates. Could you take a look at the failing recipes and see if there are differences in coordinate metadata? If that is the case, we should homogenize that instead of the data points.

jlenh commented:
Indeed! As it looks like so far and following what was identified in #2820 , the errors that I got during recipe testing can be traced back to metadata differences:

- `recipe_ipccwg1ar6ch3_fig_3_9.yml` (and `recipe_ipccwg1ar6ch3_atmosphere.yml`):
    - latitude coordinates have differing `long_name` attributes for the EC-Earth model (`latitude` vs `Latitude`)
        - cube `latitude` coordinate = `DimCoordMetadata(standard_name=latitude, long_name=latitude, var_name=lat, units=degrees, climatological=False, circular=False)`
        - ancillary `latitude` coordinate = `DimCoordMetadata(standard_name=latitude, long_name=Latitude, var_name=lat, units=degrees, climatological=False, circular=False)`
    - The modified preprocessor here would thus indeed work for these `ipccwg1ar6ch3` recipes as no check is done on the `long_name` attribute directly in the added fixes (only `var_name` and `standard_name`)
    - Should a dataset fix be included in https://github.com/ESMValGroup/ESMValCore/blob/main/esmvalcore/cmor/_fixes/cmip6/ec_earth3.py ?

- `recipe_russell18jgr.yml`:
    - Missing attribute in `latitude`/`longitude` grid coordinates between variable files for `/cmip5/output1/CNRM-CERFACS/CNRM-CM5/historical/mon/ocnBgchem/Omon/r1i1p1/v20120731/fgco2_Omon_CNRM-CM5_historical_r1i1p1_*.nc` and `CNRM-CERFACS/CNRM-CM5/historical/fx/ocean/fx/r0i0p0/v20130826/sftof/sftof_fx_CNRM-CM5_historical_r0i0p0.nc`
        - cube coordinates = `CoordMetadata(standard_name='latitude', long_name='latitude coordinate', var_name='lat', units=Unit('degrees'), attributes={'_CoordinateAxisType': 'Lat'}, coord_system=None, climatological=False)` (same for `longitude`)
        - ancillary cube coordinates = `CoordMetadata(standard_name='latitude', long_name='latitude coordinate', var_name='lat', units=Unit('degrees'), attributes={}, coord_system=None, climatological=False)` (same for `longitude`)
    - The fix here basically allows to match the cubes on the auxiliary coordinates `latitude`/`longitude` without comparing the `attributes` in the metadata (everything matching otherwise).
    - Should a dataset fix be included in https://github.com/ESMValGroup/ESMValCore/blob/main/esmvalcore/cmor/_fixes/cmip5/cnrm_cm5.py?

On a separate note, when running some of these recipes, the added fixes from the preprocessor done here seem to be used in some other cases following log messages in the `main_log_debug.txt` respective files.

jlenh commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +104,152 @@ def add_cell_measure(
+            coord_.points = (
+                coord_.core_points().astype(np.float32).astype(np.float64)
+            )
+            cube_coord_ = cube_coord.copy()
+            if np.all(coord_.points == cube_coord_.points):
+                cube_dims = cube.coord_dims(cube_coord_)
+                msg = (
+                    f"Found a matching casted coordinate for {coord_.var_name}"
+                    f" with coordinate {coord_.var_name}"
+                    f" in the cube of variable '{cube.var_name}'."
+                )
+                logger.debug(msg)
+                break
+            # Trying to find a "close" coordinate for 1D cases
+            if cube_coord_.points.ndim == 1:
+                cube_coord_.points = (
+                    cube_coord_.core_points()
+                    .astype(np.float32)
+                    .astype(np.float64)
+                )
+                acceptable_relative_difference = 0.1
+                min_diff = np.abs(
+                    np.diff(
+                        cube_coord_.core_points(),
+                    ),
+                ).min()
+                atol = acceptable_relative_difference * min_diff
+                if np.allclose(
+                    coord_.points,
+                    cube_coord_.core_points(),
+                    rtol=0,
+                    atol=atol,
+                ):
+                    cube_dims = cube.coord_dims(cube_coord_)
+                    msg = (
+                        f"Found a close coordinate for {coord_.var_name}"
+                        f" with coordinate {coord_.var_name}"
+                        f" in the cube of variable '{cube.var_name}'."
+                    )
+                    logger.debug(msg)
```
Running the aforementioned recipes with the simpler code worked all good (`recipe_ipccwg1ar6ch3_fig_3_9.yml` not finished but the preprocessor has not returned any errors so far when executed).

I think it is the sanest way to go forward regarding how iris works with coordinates without making it overly complicated for nothing! Thanks a lot for the help on this Bouwe!

jlenh commented on esmvalcore/preprocessor/_supplementary_vars.py:
```diff
@@ -103,6 +103,120 @@ def add_cell_measure(
+    else:
+        data_dims = [None] * ancillary.ndim
+        for coord in ancillary.coords():
+            try:
+                for ancillary_dim, cube_dim in zip(
+                    ancillary.coord_dims(coord),
+                    cube.coord_dims(coord),
+                    strict=True,
+                ):
+                    data_dims[ancillary_dim] = cube_dim
+            except iris.exceptions.CoordinateNotFoundError:
+                cube_dims = find_matching_coord(
+                    coord,
+                    cube,
+                )
+                if cube_dims is not None:
+                    for ancillary_dim, cube_dim in zip(
+                        ancillary.coord_dims(coord),
+                        cube_dims,
+                        strict=True,
+                    ):
+                        data_dims[ancillary_dim] = cube_dim
```
You are completely right, did not think about that!

jlenh commented:
> Thanks @jlenh! I left a few more suggestions to improve the readability of the code, but this already looks good.

Thanks a lot for taking another look at it! I have implemented your comments :)
